[OSDOCS-10694]: OSD-GCP - Document GCP Workload Identity Federation (#76600)
openshift/openshift-docs: modules/create-wif-cluster-cli.adoc

// Module included in the following assemblies:
//
// * osd_install_access_delete_cluster/creating-a-gcp-cluster-with-workload-identity-federation.adoc

:_mod-docs-content-type: PROCEDURE
[id="create-wif-cluster-cli_{context}"]
= Creating a Workload Identity Federation cluster using the OCM CLI

You can create an {product-title} on {GCP} cluster with Workload Identity Federation (WIF) using the OpenShift Cluster Manager CLI (`ocm`) in interactive or non-interactive mode.

> Before creating the cluster, you must first create a WIF configuration. You can create a WIF configuration in auto or manual mode.

Reviewer: ```suggestion
Before creating the cluster, you must first create a WIF configuration. 
```
You talk about manual and auto mode in the procedure so I don't think you need to mention it here too.

[NOTE]
====
Migrating an existing non-WIF cluster to a WIF configuration is not supported. This feature can only be enabled during new cluster creation.
====

[id="create-wif-configuration_{context}"]
== Creating a WIF configuration

.Procedure
You can create a WIF configuration using auto mode or manual mode.

> **Auto mode** enables you to automatically create the service accounts for {product-title} components as well as other IAM resources.

Reviewer: ```suggestion
_Auto mode_ enables you to automatically create the service accounts for {product-title} components as well as other IAM resources.
```
As per SSG, we can't use bold for emphasis but can italicize.

> Alternatively, you can use **Manual mode* to review the details in regard to the service accounts for {product-title} components and other IAM resources before using your preferred options.

Reviewer: ```suggestion
Alternatively, you can use _Manual mode_ to review the details in regard to the service accounts for {product-title} components and other IAM resources before using your preferred options.
```

* Based on your preference, run one of the following commands to create a WIF configuration:

** Create a WIF configuration in auto mode by running the following command:
+
[source,terminal]
----
$ ocm gcp create wif-config --name <wif_name> \ <1>
  --project <gcp_project_id> \ <2>
  --tag-enforcement=<true|false> \ <3>
  --versions <required_versions> <4>

----
<1> Replace `<wif_name>` with the name of your WIF configuration.

> <2> Replace `<gcp_project_id>` with the ID  of the {GCP} project where the WIF configuration will be implemented.

Reviewer: ```suggestion
<2> Replace `<gcp_project_id>` with the ID of the {GCP} project where the WIF configuration will be implemented.
```

<3> Replace`<true|false>` with `false`.

> <4> Optional: Replace `<required_versions>` with the list of Openshift versions that the WIF configuration will need to support. If no value is provided, the latest available version will be supported by default.

Reviewer: ```suggestion
<4> Optional: Replace `<required_versions>` with the list of Openshift versions that the WIF configuration needs to support. If no value is provided, the latest available version is supported by default.
```

2024/09/26 13:05:48 IAM service account osd-worker-oeub created
2024/09/26 13:05:49 IAM service account osd-control-plane-oeub created
2024/09/26 13:05:49 IAM service account openshift-gcp-ccm-oeub created
2024/09/26 13:05:50 IAM service account openshift-gcp-pd-csi-driv-oeub created
2024/09/26 13:05:50 IAM service account openshift-image-registry-oeub created
2024/09/26 13:05:51 IAM service account openshift-machine-api-gcp-oeub created
2024/09/26 13:05:51 IAM service account osd-deployer-oeub created
2024/09/26 13:05:52 IAM service account cloud-credential-operator-oeub created
2024/09/26 13:05:52 IAM service account openshift-cloud-network-c-oeub created
2024/09/26 13:05:53 IAM service account openshift-ingress-gcp-oeub created
2024/09/26 13:05:55 Role "osd_deployer_v4.17" updated
----
--
+
** Create a WIF configuration in manual mode by running the following command:
+
[source,terminal]
----
$ ocm gcp create wif-config --name <wif_name> \ <1>
  --project <gcp_project_id> \ <2>
  --manual-mode
----
<1> Replace `<wif_name>` with the name of your WIF configuration.
<2> Replace `<gcp_project_id>` with the ID  of the {GCP} project where the WIF configuration will be implemented.
+

> Once the WIF is configured, the following service accounts, group and roles are created.

Reviewer: ```suggestion
Once the WIF is configured, the following service accounts, groups, and roles are created.
```

|===

|Service Account/Group
|GCP pre-defined roles and Red Hat custom roles

|osd-deployer
|osd_deployer_v4.17

|osd-control-plane
|- compute.instanceAdmin
- compute.networkAdmin
- compute.securityAdmin
- compute.storageAdmin

|osd-worker
|- compute.storageAdmin
- compute.viewer

|cloud-credential-operator-gcp-ro-creds
|cloud_credential_operator_gcp_ro_creds_v4.17

|openshift-cloud-network-config-controller-gcp
|openshift_cloud_network_config_controller_gcp_v4.17

|openshift-gcp-ccm
|openshift_gcp_ccm_v4.17

|openshift-gcp-pd-csi-driver-operator
|- compute.storageAdmin
- iam.serviceAccountUser
- resourcemanager.tagUser
- openshift_gcp_pd_csi_driver_operator_v4.17

|openshift-image-registry-gcp
|openshift_image_registry_gcs_v4.17

|openshift-ingress-gcp
|openshift_ingress_gcp_v4.17

|openshift-machine-api-gcp
|openshift_machine_api_gcp_v4.17

|Access via SRE group:sd-sre-platform-gcp-access
|sre_managed_support

|===

For further details about WIF configuration roles and their assigned permissions, see link:https://github.com/openshift/managed-cluster-config/blob/master/resources/wif/4.17/vanilla.yaml[managed-cluster-config].

[id="create-wif-cluster_{context}"]

> == Creating a WIF cluster

Reviewer: ```suggestion
== Creating a WIF cluster using the OCM CLI
```
Slight repetition but I don't think it's a strong enough title as it is.

.Procedure

* Run one of the following commands to create an {product-title} on (GCP) cluster with WIF configuration:

** Create a cluster in interactive mode by running the following command:
+
[source,terminal]
----
ocm create cluster --interactive <1>
----
<1> `interactive` mode enables you to specify configuration options at the interactive prompts.
+
** Create a cluster in non-interactive mode by running the following command:
+
--
[source,terminal]
----

> ocm create cluster <cluster_name> \ <1>

Reviewer: Missing the `$`

--provider=gcp \
--wif-config <wif_name> \ <2>
--region <gcp_region> \ <3>
--authentication-type <service-account|wif>  \ <4>
--compute-machine-type <instance_type> \ <5>
--version <version> <6>
----
<1> Replace `<cluster_name>` with a name for your cluster.
<2> Replace `<wif_name>` with the name of the newly created or pre-existing WIF configuration.
<3> Replace `<gcp_region>` with the {GCP} region where the new cluster will be deployed.
<4> Select "wif" as your authentication type.
<5> Optional: Replace `<instance_type>` with the instance type you want for your cluster.
<6> Optional: Replace `<version>` with {product-title} version you desire for your cluster.
--
+
[NOTE]
====

> `--gcp-tag-enforcement` cannot  be set to `true` if `service-account` is specified as the authentication type.

Reviewer: ```suggestion
`--gcp-tag-enforcement` cannot be set to `true` if `service-account` is specified as the authentication type.
```

====

[id="wif-configuration-update_{context}"]
== Updating a WIF configuration

[NOTE]
====
Updating a WIF configuration is only applicable for y-stream updates.
====

You can update a WIF configuration to the latest {product-title} version by running the following command:

--
[source,terminal]
----

> ocm gcp update wif-config --version <version> \ <1>

Reviewer: Missing the `$`

--name <wif_name> <2>
----
<1> Replace `<version>` with the latest Openshift version.
<2> Replace `<wif_name>` with the name of the WIF configuration you want to update.

> --

Reviewer: Again, what about deleting a WIF configuration?

|openshift-image-registry-gcp
|openshift_image_registry_gcs_v4.17

|openshift-ingress-gcp
|openshift_ingress_gcp_v4.17

|openshift-machine-api-gcp
|openshift_machine_api_gcp_v4.17

|Access via SRE group:sd-sre-platform-gcp-access
|sre_managed_support

|===

For further details about WIF configuration roles and their assigned permissions, see link:https://github.com/openshift/managed-cluster-config/blob/master/resources/wif/4.17/vanilla.yaml[managed-cluster-config].

[id="create-wif-cluster_{context}"]
== Creating a WIF cluster

.Procedure
You can create a WIF cluster using the `interactive` mode or the `non-interactive` mode.

In `interactive` mode, attributes are displayed automatically as prompts during the creation of the cluster, and you specify the values for those prompts based on specified requirements.

In `non-interactive` mode, you specify the values for specific attributes within the command.

* Based on your mode preference, run one of the following commands to create an {product-title} on (GCP) cluster with WIF configuration:

** Create a cluster in interactive mode by running the following command:
+
[source,terminal]
----

> ocm create cluster --interactive <1>

Reviewer: Hi @mletalie, would it be worth following the ROSA CLI guide example here, and having an example output to explain the various CLI fields? Be more work I know. 
![image](https://github.com/user-attachments/assets/904d0b04-6b85-469d-a080-a947ec9a7117)

// Module included in the following assemblies:
//
// * osd_install_access_delete_cluster/creating-a-gcp-cluster-with-workload-identity-federation.adoc

:_mod-docs-content-type: PROCEDURE
[id="create-wif-cluster-cli_{context}"]
= Creating a Workload Identity Federation cluster using the OCM CLI

You can create an {product-title} on {GCP} cluster with Workload Identity Federation (WIF) using the OpenShift Cluster Manager CLI (`ocm`) in interactive or non-interactive mode.

[IMPORTANT]
====
To create a WIF-enabled cluster, the OpenShift Cluster Manager CLI (`ocm`) must be version 1.0.2 or greater.
====

Before creating the cluster, you must first create a WIF configuration.
[NOTE]
====
Migrating an existing non-WIF cluster to a WIF configuration is not supported. This feature can only be enabled during new cluster creation.
====

[id="create-wif-configuration_{context}"]
== Creating a WIF configuration

.Procedure
You can create a WIF configuration using the `auto` mode or the `manual` mode.

The `auto` mode enables you to automatically create the service accounts for {product-title} components as well as other IAM resources.

> Alternatively, you can use the `manual` mode to create a `script.sh` file in order review the details in regard to the service accounts for {product-title} components and other IAM resources before using your preferred options.

Reviewer 2: In manual mode user is provided with commands (script.sh) and user needs to manually configure their cloud.
